There is a sharp increase in the number of test errors, it seems the error is usually a NPE:

<details>
<summary>stack trace</summary>
```
java.lang.IllegalStateException: java.lang.NullPointerException: Cannot invoke "com.sun.source.util.TreePath.getLeaf()" because "diagnosicPath" is null
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.parse(JavacTaskImpl.java:252)
	at org.eclipse.jdt.core.dom.JavacCompilationUnitResolver.parse(JavacCompilationUnitResolver.java:511)
	at org.eclipse.jdt.core.dom.JavacCompilationUnitResolver.toCompilationUnit(JavacCompilationUnitResolver.java:427)
	at org.eclipse.jdt.core.dom.ASTParser.internalCreateASTCached(ASTParser.java:1253)
	at org.eclipse.jdt.core.dom.ASTParser.lambda$0(ASTParser.java:1132)
	at org.eclipse.jdt.internal.core.JavaModelManager.cacheZipFiles(JavaModelManager.java:5762)
	at org.eclipse.jdt.core.dom.ASTParser.internalCreateAST(ASTParser.java:1132)
	at org.eclipse.jdt.core.dom.ASTParser.createAST(ASTParser.java:880)
	at org.eclipse.jdt.internal.core.CompilationUnit.buildStructure(CompilationUnit.java:218)
	at org.eclipse.jdt.internal.core.Openable.generateInfos(Openable.java:245)
	at org.eclipse.jdt.internal.core.JavaElement.openWhenClosed(JavaElement.java:585)
	at org.eclipse.jdt.internal.core.JavaElement.getElementInfo(JavaElement.java:308)
	at org.eclipse.jdt.internal.core.JavaElement.getElementInfo(JavaElement.java:294)
	at org.eclipse.jdt.internal.core.CompilationUnit.getCompilationUnitElementInfo(CompilationUnit.java:1652)
	at org.eclipse.jdt.internal.core.CompilationUnit.getCustomOptions(CompilationUnit.java:1630)
	at org.eclipse.jdt.core.ICompilationUnit.getOptions(ICompilationUnit.java:607)
	at org.eclipse.jdt.core.dom.ASTParser.setSource(ASTParser.java:624)
	at org.eclipse.jdt.core.tests.dom.ConverterTestSetup.runConversion(ConverterTestSetup.java:423)
	at org.eclipse.jdt.core.tests.dom.ConverterTestSetup.runConversion(ConverterTestSetup.java:413)
	at org.eclipse.jdt.core.tests.dom.ConverterTestSetup.runConversion(ConverterTestSetup.java:409)
	at org.eclipse.jdt.core.tests.dom.ConverterTestSetup.runConversion(ConverterTestSetup.java:389)
	at org.eclipse.jdt.core.tests.dom.ASTConverterTest.test0153(ASTConverterTest.java:3450)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at junit.framework.TestCase.runTest(TestCase.java:177)
	at org.eclipse.jdt.core.tests.junit.extension.TestCase.runTest(TestCase.java:972)
	at junit.framework.TestCase.runBare(TestCase.java:142)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:130)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at org.eclipse.jdt.core.tests.model.SuiteOfTestCases$Suite.runTest(SuiteOfTestCases.java:114)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.eclipse.jdt.core.tests.model.SuiteOfTestCases$Suite.superRun(SuiteOfTestCases.java:98)
	at org.eclipse.jdt.core.tests.model.SuiteOfTestCases$Suite$1.protect(SuiteOfTestCases.java:86)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at org.eclipse.jdt.core.tests.model.SuiteOfTestCases$Suite.run(SuiteOfTestCases.java:95)
	at junit.framework.TestSuite.runTest(TestSuite.java:241)
	at junit.framework.TestSuite.run(TestSuite.java:236)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:90)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:757)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.pde.internal.junit.runtime.RemotePluginTestRunner.main(RemotePluginTestRunner.java:83)
	at org.eclipse.pde.internal.junit.runtime.CoreTestApplication.start(CoreTestApplication.java:28)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:208)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:143)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:109)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:439)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:271)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:668)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:605)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1481)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1454)
Caused by: java.lang.NullPointerException: Cannot invoke "com.sun.source.util.TreePath.getLeaf()" because "diagnosicPath" is null
	at org.eclipse.jdt.internal.javac.JavacProblemConverter.getDiagnosticPosition(JavacProblemConverter.java:116)
	at org.eclipse.jdt.internal.javac.JavacProblemConverter.createJavacProblem(JavacProblemConverter.java:87)
	at org.eclipse.jdt.core.dom.JavacCompilationUnitResolver.lambda$16(JavacCompilationUnitResolver.java:458)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at org.eclipse.jdt.core.dom.JavacCompilationUnitResolver.lambda$15(JavacCompilationUnitResolver.java:457)
	at jdk.compiler/com.sun.tools.javac.util.Log.writeDiagnostic(Log.java:730)
	at jdk.compiler/com.sun.tools.javac.util.Log$DefaultDiagnosticHandler.report(Log.java:711)
	at jdk.compiler/com.sun.tools.javac.util.Log.report(Log.java:664)
	at jdk.compiler/com.sun.tools.javac.util.AbstractLog.error(AbstractLog.java:126)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.reportSyntaxError(JavacParser.java:485)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.reportSyntaxError(JavacParser.java:472)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.accept(JavacParser.java:513)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.accept(JavacParser.java:502)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.localVariableDeclarations(JavacParser.java:2946)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.blockStatement(JavacParser.java:2931)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.blockStatements(JavacParser.java:2762)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.block(JavacParser.java:2732)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.block(JavacParser.java:2746)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.methodDeclaratorRest(JavacParser.java:4910)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.constructorOrMethodOrFieldDeclaration(JavacParser.java:4707)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.classOrInterfaceOrRecordBodyDeclaration(JavacParser.java:4632)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.classInterfaceOrRecordBody(JavacParser.java:4571)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.classDeclaration(JavacParser.java:4271)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.classOrRecordOrInterfaceOrEnumDeclaration(JavacParser.java:4218)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.typeDeclaration(JavacParser.java:4207)
	at jdk.compiler/com.sun.tools.javac.parser.JavacParser.parseCompilationUnit(JavacParser.java:4014)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.parse(JavaCompiler.java:652)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.parse(JavaCompiler.java:629)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.parse(JavaCompiler.java:689)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.parseFiles(JavaCompiler.java:1043)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler$InitialFileParser.parse(JavaCompiler.java:1980)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.parseFiles(JavaCompiler.java:1030)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.parseInternal(JavacTaskImpl.java:258)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.invocationHelper(JavacTaskImpl.java:152)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.parse(JavacTaskImpl.java:248)
	... 58 more


```
</details>

--- a/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacCompilationUnitResolver.java
+++ b/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/core/dom/JavacCompilationUnitResolver.java
@@ -20,7 +20,6 @@ import java.util.Collection;
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.HashMap;
-import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
@@ -68,7 +67,10 @@ import org.eclipse.jdt.internal.javac.JavacProblemConverter;
 import org.eclipse.jdt.internal.javac.JavacUtils;
 
 import com.sun.source.util.JavacTask;
+import com.sun.source.util.TaskEvent;
+import com.sun.source.util.TaskListener;
 import com.sun.tools.javac.api.JavacTool;
+import com.sun.tools.javac.api.MultiTaskListener;
 import com.sun.tools.javac.file.JavacFileManager;
 import com.sun.tools.javac.parser.JavadocTokenizer;
 import com.sun.tools.javac.parser.Scanner;
@@ -382,7 +384,7 @@ public class JavacCompilationUnitResolver implements ICompilationUnitResolver {
 			}
 		}
 		if (!unit.types().isEmpty()) {
-			List types = unit.types();
+			List<AbstractTypeDeclaration> types = unit.types();
 			for( int i = 0; i < types.size(); i++ ) {
 				ITypeBinding tb = ((AbstractTypeDeclaration) types.get(i)).resolveBinding();
 				if (tb != null) {
@@ -462,6 +464,14 @@ public class JavacCompilationUnitResolver implements ICompilationUnitResolver {
 				}
 			});
 		};
+		MultiTaskListener.instance(context).add(new TaskListener() {
+			@Override
+			public void finished(TaskEvent e) {
+				if (e.getCompilationUnit() instanceof JCCompilationUnit u) {
+					problemConverter.registerUnit(e.getSourceFile(), u);
+				}
+			}
+		});
 		// must be 1st thing added to context
 		context.put(DiagnosticListener.class, diagnosticListener);
 		boolean docEnabled = JavaCore.ENABLED.equals(compilerOptions.get(JavaCore.COMPILER_DOC_COMMENT_SUPPORT));

--- a/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java
+++ b/org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/javac/JavacProblemConverter.java
@@ -14,6 +14,7 @@
 package org.eclipse.jdt.internal.javac;
 
 import java.io.IOException;
+import java.util.HashMap;
 import java.util.Locale;
 import java.util.Map;
 import java.util.stream.Stream;
@@ -27,6 +28,9 @@ import org.eclipse.jdt.internal.compiler.impl.CompilerOptions;
 import org.eclipse.jdt.internal.compiler.problem.ProblemReporter;
 import org.eclipse.jdt.internal.compiler.problem.ProblemSeverities;
 
+import com.sun.source.tree.Tree;
+import com.sun.source.util.TreePath;
+import com.sun.tools.javac.api.JavacTrees;
 import com.sun.tools.javac.code.Kinds;
 import com.sun.tools.javac.code.Kinds.KindName;
 import com.sun.tools.javac.code.Symbol;
@@ -39,8 +43,11 @@ import com.sun.tools.javac.parser.Tokens.TokenKind;
 import com.sun.tools.javac.tree.JCTree;
 import com.sun.tools.javac.tree.JCTree.JCBlock;
 import com.sun.tools.javac.tree.JCTree.JCClassDecl;
+import com.sun.tools.javac.tree.JCTree.JCCompilationUnit;
 import com.sun.tools.javac.tree.JCTree.JCFieldAccess;
+import com.sun.tools.javac.tree.JCTree.JCIdent;
 import com.sun.tools.javac.tree.JCTree.JCMethodDecl;
+import com.sun.tools.javac.tree.JCTree.JCMethodInvocation;
 import com.sun.tools.javac.tree.JCTree.JCVariableDecl;
 import com.sun.tools.javac.util.Context;
 import com.sun.tools.javac.util.DiagnosticSource;
@@ -55,6 +62,7 @@ public class JavacProblemConverter {
 	private static final String COMPILER_WARN_MISSING_SVUID = "compiler.warn.missing.SVUID";
 	private final CompilerOptions compilerOptions;
 	private final Context context;
+	private final Map<JavaFileObject, JCCompilationUnit> units = new HashMap<>();
 
 	public JavacProblemConverter(Map<String, String> options, Context context) {
 		this(new CompilerOptions(options), context);
@@ -79,7 +87,7 @@ public class JavacProblemConverter {
 		if (severity == ProblemSeverities.Ignore || severity == ProblemSeverities.Optional) {
 			return null;
 		}
-		org.eclipse.jface.text.Position diagnosticPosition = getDiagnosticPosition(diagnostic, context);
+		org.eclipse.jface.text.Position diagnosticPosition = getDiagnosticPosition(diagnostic, context, problemId);
 		String[] arguments = getDiagnosticStringArguments(diagnostic);
 		return new JavacProblem(
 				diagnostic.getSource().getName().toCharArray(),
@@ -94,29 +102,51 @@ public class JavacProblemConverter {
 				(int) diagnostic.getColumnNumber());
 	}
 
-	private org.eclipse.jface.text.Position getDiagnosticPosition(Diagnostic<? extends JavaFileObject> diagnostic, Context context) {
+	private org.eclipse.jface.text.Position getDiagnosticPosition(Diagnostic<? extends JavaFileObject> diagnostic, Context context, int problemId) {
 		if (diagnostic.getCode().contains(".dc")) { //javadoc
 			return getDefaultPosition(diagnostic);
 		}
 		if (diagnostic instanceof JCDiagnostic jcDiagnostic) {
-			switch (jcDiagnostic.getDiagnosticPosition()) {
-				case JCClassDecl jcClassDecl: return getDiagnosticPosition(jcDiagnostic, jcClassDecl);
-				case JCVariableDecl jcVariableDecl: return getDiagnosticPosition(jcDiagnostic, jcVariableDecl);
-				case JCMethodDecl jcMethodDecl: return getDiagnosticPosition(jcDiagnostic, jcMethodDecl);
-				case JCFieldAccess jcFieldAccess:
-					if (getDiagnosticArgumentByType(jcDiagnostic, KindName.class) != KindName.PACKAGE && getDiagnosticArgumentByType(jcDiagnostic, Symbol.PackageSymbol.class) == null) {
-						// TODO here, instead of recomputing a position, get the JDT DOM node and call the Name (which has a position)
-						return new org.eclipse.jface.text.Position(jcFieldAccess.getPreferredPosition() + 1, jcFieldAccess.getIdentifier().length());
+			TreePath diagnosticPath = getTreePath(jcDiagnostic);
+			if (problemId == IProblem.ParameterMismatch && diagnosticPath != null && !(diagnosticPath.getLeaf() instanceof JCMethodInvocation)) {
+				diagnosticPath = diagnosticPath.getParentPath();
+				if (diagnosticPath.getLeaf() instanceof JCMethodInvocation method) {
+					var selectExpr = method.getMethodSelect();
+					if (selectExpr instanceof JCIdent methodNameIdent) {
+						int start = methodNameIdent.getStartPosition();
+						int end = methodNameIdent.getEndPosition(this.units.get(jcDiagnostic.getSource()).endPositions);
+						return new org.eclipse.jface.text.Position(start, end - start);
 					}
-					// else: fail-through
-				default:
-					org.eclipse.jface.text.Position result = getMissingReturnMethodDiagnostic(jcDiagnostic, context);
-					if (result != null) {
-						return result;
+					if (selectExpr instanceof JCFieldAccess methodFieldAccess) {
+						int start = methodFieldAccess.getPreferredPosition() + 1; // after dot
+						int end = methodFieldAccess.getEndPosition(this.units.get(jcDiagnostic.getSource()).endPositions);
+						return new org.eclipse.jface.text.Position(start, end - start);
 					}
-					if (jcDiagnostic.getStartPosition() == jcDiagnostic.getEndPosition()) {
-						return getPositionUsingScanner(jcDiagnostic, context);
-					}
+				}
+			}
+			Tree element = diagnosticPath != null ? diagnosticPath.getLeaf() :
+				jcDiagnostic.getDiagnosticPosition() instanceof Tree tree ? tree :
+				null;
+			if (element != null) {
+				switch (element) {
+					case JCClassDecl jcClassDecl: return getDiagnosticPosition(jcDiagnostic, jcClassDecl);
+					case JCVariableDecl jcVariableDecl: return getDiagnosticPosition(jcDiagnostic, jcVariableDecl);
+					case JCMethodDecl jcMethodDecl: return getDiagnosticPosition(jcDiagnostic, jcMethodDecl);
+					case JCFieldAccess jcFieldAccess:
+						if (getDiagnosticArgumentByType(jcDiagnostic, KindName.class) != KindName.PACKAGE && getDiagnosticArgumentByType(jcDiagnostic, Symbol.PackageSymbol.class) == null) {
+							// TODO here, instead of recomputing a position, get the JDT DOM node and call the Name (which has a position)
+							return new org.eclipse.jface.text.Position(jcFieldAccess.getPreferredPosition() + 1, jcFieldAccess.getIdentifier().length());
+						}
+						// else: fail-through
+					default:
+						org.eclipse.jface.text.Position result = getMissingReturnMethodDiagnostic(jcDiagnostic, context);
+						if (result != null) {
+							return result;
+						}
+						if (jcDiagnostic.getStartPosition() == jcDiagnostic.getEndPosition()) {
+							return getPositionUsingScanner(jcDiagnostic, context);
+						}
+				}
 			}
 		}
 		return getDefaultPosition(diagnostic);
@@ -551,7 +581,26 @@ public class JavacProblemConverter {
 				return IProblem.ShouldReturnValue;
 			}
 		}
+		if (diagnostic instanceof JCDiagnostic jcDiagnostic && jcDiagnostic.getDiagnosticPosition() instanceof JCTree tree) {
+			JCCompilationUnit unit = units.get(jcDiagnostic.getSource());
+			if (unit != null) {
+				TreePath path = JavacTrees.instance(context).getPath(unit, tree);
+				if (path.getParentPath().getLeaf() instanceof JCMethodInvocation) {
+					return IProblem.ParameterMismatch;
+				}
+			}
+		}
 		return IProblem.TypeMismatch;
+	}
+
+	private TreePath getTreePath(Diagnostic<?> diagnostic) {
+		if (diagnostic instanceof JCDiagnostic jcDiagnostic && jcDiagnostic.getDiagnosticPosition() instanceof JCTree tree) {
+			JCCompilationUnit unit = units.get(jcDiagnostic.getSource());
+			if (unit != null) {
+				return JavacTrees.instance(context).getPath(unit, tree);
+			}
+		}
+		return null;
 	}
 
 	private int convertNotVisibleAccess(Diagnostic<?> diagnostic) {
@@ -586,5 +635,9 @@ public class JavacProblemConverter {
 			case METHOD -> IProblem.AmbiguousMethod;
 			default -> 0;
 		};
+	}
+
+	public void registerUnit(JavaFileObject javaFileObject, JCCompilationUnit unit) {
+		this.units.put(javaFileObject, unit);
 	}
 }

